src/consensus/mod.rs

```diff
@@ -1,58 +1,191 @@
-            proto::RequestKind::Consensus(_) => self.paxos.handle_phase_1a(request).await,
-            proto::RequestKind::Consensus(_) => self.paxos.handle_phase_1b(request).await,
-            proto::RequestKind::Consensus(_) => self.paxos.handle_phase_2a(request).await,
-            proto::RequestKind::Consensus(_) => self.paxos.handle_phase_2b(request).await,
+            proto::RequestKind::Consensus(req_type) => {
+                if self.paxos.is_some() {
+                    match req_type {
+                        proto::Consensus::FastRoundPhase2bMessage(msg) => {
+                            self.handle_fast_round(msg).await
+                        }
+                        // TODO: need to figure out communication b/w paxos and fast paxos
+                        _ => unimplemented!(),
+                    }
+                } else {
+                    Err(Error::new_unexpected_request(None))
+                }
+            }
             _ => Err(Error::new_unexpected_request(None)),
         }
     }
 
-    async fn handle_fast_round(&self, request: Request) -> crate::Result<Response> {
-        unimplemented!()
+    async fn handle_fast_round<'b>(
+        &'b mut self,
+        request: &'b proto::FastRoundPhase2bMessage,
+    ) -> crate::Result<Vec<Endpoint>> {
+        if request.config_id != self.config_id {
+            return Err(Error::new_unexpected_request(None));
+        }
+
+        if self.votes_received.contains(&request.sender) {
+            return Err(Error::vote_already_received());
+        }
+
+        if (self.decided.load(Ordering::SeqCst)) {
+            return Err(Error::already_reached_consensus());
+        }
+
+        self.votes_received.insert(request.sender.clone());
+
+        let count = self
+            .votes_per_proposal
+            .entry(request.endpoints.clone())
+            .and_modify(|votes| *votes += 1)
+            .or_insert(0);
+
+        let F = ((self.size - 1) as f64 / 4f64).floor();
+
+        if (self.votes_received.len() >= (self.size as f64 - F) as usize) {
+            if (*count >= (self.size as f64 - F) as usize) {
+                return self.on_decide(request.endpoints.clone());
+            }
+        }
+
+        Err(Error::fast_round_failure())
+    }
+
+    fn on_decide(&mut self, hosts: Vec<Endpoint>) -> Result<Vec<Endpoint>> {
+        // This is the only place where the value is set to `true`.
+        self.decided.store(true, Ordering::SeqCst);
+
+        let paxos_instance = self.paxos.take();
+
+        if let Some(instance) = paxos_instance {
+            drop(instance);
+        }
+
+        Ok(hosts)
+    }
+
+    async fn start_classic_round(&mut self) -> Result<()> {
+        if !self.decided.load(Ordering::SeqCst) {
+            if let Some(paxos) = &mut self.paxos {
+                paxos.start_round().await?;
+            }
+        }
+        Ok(())
+    }
+
+    fn get_random_delay(&self) -> Duration {
```

Comment:
could put this in `src/common.rs`

Reply:
will it actually be used anywhere outside of this module though?

Reply:
maybe 🤷‍♂